api/auth.proto

```diff
@@ -0,0 +1,43 @@
+syntax = "proto3";
+
+package envoy.api.v2;
+
+message AuthConfig {
+
+  // Check peer identity using X.509 certificate
+  message X509 {
+    enum VerifyType {
+      option allow_alias = true;
+      DEFAULT = 0;
+      ALL = 0;
+      ANY = 1;
+    }
+
+    // should all/any condition(s) matches
+    VerifyType verify_type = 1;
+
+    // NOTE: another level of indirect to get has_XXX behaviour
+
+    // the sha256 hash of the certificate
+    message Sha256HashesOptional {
+      repeated string sha256_hashes = 1;
+    }
+    Sha256HashesOptional sha256_hashes = 2;
```

Comment:
What about having a list of accepted identities, each identity object has SHA256, Subject, SANs?

Reply:
+1, we should refactor this so that we have a repeated oneof, such that we can compose multiple types of auth. 

I also think that we should handle both white list (only allow if matches all of) and black list (do not allow if matches all of).

Reply:
great idea, i copied the current configs from *ssl_context* and didn't think much about this.

i think we can have a list of x509 config, whose fields(sha256/subject/sans) are all optional, and if present,  the peer certificate should match all fields to be considered a successful match.

and have another extra type to specify wheather we are doing while-list or black-list.